src/solvers/dgsem_tree/containers_2d.jl

```diff
@@ -1255,6 +1255,125 @@ function init_mpi_mortars!(mpi_mortars, elements, mesh::TreeMesh2D)
+  antidiffusive_flux2 = unsafe_wrap(Array, pointer(_antidiffusive_flux2),
+                                    (n_variables, n_nodes, n_nodes+1, capacity))
+
+  return ContainerAntidiffusiveFlux2D{uEltype}(antidiffusive_flux1, antidiffusive_flux2,
+                                               _antidiffusive_flux1, _antidiffusive_flux2)
+end
+
+nvariables(fluxes::ContainerAntidiffusiveFlux2D) = size(fluxes.antidiffusive_flux1, 1)
+nnodes(fluxes::ContainerAntidiffusiveFlux2D) = size(fluxes.antidiffusive_flux1, 3)
+
+# Only one-dimensional `Array`s are `resize!`able in Julia.
+# Hence, we use `Vector`s as internal storage and `resize!`
+# them whenever needed. Then, we reuse the same memory by
+# `unsafe_wrap`ping multi-dimensional `Array`s around the
+# internal storage.
+function Base.resize!(fluxes::ContainerAntidiffusiveFlux2D, capacity)
+  n_nodes = nnodes(fluxes)
+  n_variables = nvariables(fluxes)
+
+  @unpack _antidiffusive_flux1, _antidiffusive_flux2 = fluxes
+
+  resize!(_antidiffusive_flux1, n_variables * (n_nodes+1) * n_nodes * capacity)
+  fluxes.antidiffusive_flux1 = unsafe_wrap(Array, pointer(_antidiffusive_flux1),
+                                           (n_variables, n_nodes+1, n_nodes, capacity))
+  resize!(_antidiffusive_flux2, n_variables * n_nodes * (n_nodes+1) * capacity)
+  fluxes.antidiffusive_flux2 = unsafe_wrap(Array, pointer(_antidiffusive_flux2),
+                                           (n_variables, n_nodes, n_nodes+1, capacity))
+
+  return nothing
+end
+
 
+# Container data structure (structure-of-arrays style) for variables used for IDP limiting
+mutable struct ContainerShockCapturingIndicatorIDP2D{uEltype<:Real}
+  alpha::Array{uEltype, 3}                  # [i, j, element]
+  alpha1::Array{uEltype, 3}
+  alpha2::Array{uEltype, 3}
+  variable_bounds::Vector{Array{uEltype, 3}}
+  # internal `resize!`able storage
+  _alpha::Vector{uEltype}
+  _alpha1::Vector{uEltype}
+  _alpha2::Vector{uEltype}
+  _variable_bounds::Vector{Vector{uEltype}}
+end
+
+function ContainerShockCapturingIndicatorIDP2D{uEltype}(capacity::Integer, n_nodes, number_bounds) where uEltype<:Real
+  nan_uEltype = convert(uEltype, NaN)
+
+  # Initialize fields with defaults
+  _alpha = fill(nan_uEltype, n_nodes * n_nodes * capacity)
+  alpha = unsafe_wrap(Array, pointer(_alpha), (n_nodes, n_nodes, capacity))
+  _alpha1 = fill(nan_uEltype, (n_nodes+1) * n_nodes * capacity)
+  alpha1 = unsafe_wrap(Array, pointer(_alpha1), (n_nodes+1, n_nodes, capacity))
+  _alpha2 = fill(nan_uEltype, n_nodes * (n_nodes+1) * capacity)
+  alpha2 = unsafe_wrap(Array, pointer(_alpha2), (n_nodes, n_nodes+1, capacity))
+
+  _variable_bounds = Vector{Vector{uEltype}}(undef, number_bounds)
+  variable_bounds  = Vector{Array{uEltype, 3}}(undef, number_bounds)
+  for i in 1:number_bounds
+    _variable_bounds[i] = fill(nan_uEltype, n_nodes * n_nodes * capacity)
+    variable_bounds[i] = unsafe_wrap(Array, pointer(_variable_bounds[i]), (n_nodes, n_nodes, capacity))
+  end
```

Comment:
I was just thinking about this again. In later PRs, we add more limiters and, therefore, have to calculate and store more bounds. We have bounds for the two-sided state limiting, the one-sided positivity limiting, and bounds for the entropy limiters as well. Since positivity bounds and one of the two-sided state limiter bounds match (if both are activated), the correct order in `variables_bounds` becomes very messy.
Rn I'm using something like [this](https://github.com/bennibolm/Trixi.jl/blob/9920815f9153873b3087a416622bb44bc80e4366/src/solvers/dgsem_tree/indicators_2d.jl#L587-L595):
```
@unpack variable_bounds = indicator.cache.ContainerShockCapturingIndicator
if Trixi.density in variables_cons && density_tvd
    if Trixi.density == variables_cons[index]
      var_min = variable_bounds[1]
    else
      var_min = variable_bounds[2 * density_tvd + spec_entropy + math_entropy + index - 1]
    end
  else
    var_min = variable_bounds[2 * density_tvd + spec_entropy + math_entropy + index]
  end
```

Would it be better to use `NamedTuple`'s instead of vectors? We could denote the entry with something like `variable_min` and access it directly with something like:
```
var_min = indicator.cache.ContainerShockCapturingIndicator.density_min
```
Or does this lead to type instabilities or problems while resizing the vectors/tuples. I'm not sure about that.

Reply:
But maybe that's more of an issue when we add the state limiting.

Reply:
If I were you, I'd address this once you actually add the additional capabilities (YAGNI for now).